src/sentry/static/sentry/app/views/organizationEventsV2/transactionView/spanTree.tsx

```diff
@@ -0,0 +1,205 @@
+import React from 'react';
+import styled from 'react-emotion';
+import {get} from 'lodash';
+
+import {t} from 'app/locale';
+
+import {SpanType, SpanChildrenLookupType, ParsedTraceType} from './types';
+import {
+  boundsGenerator,
+  SpanBoundsType,
+  SpanGeneratedBoundsType,
+  generateSpanColourPicker,
+} from './utils';
+import {DragManagerChildrenProps} from './dragManager';
+import SpanGroup from './spanGroup';
+import {SpanRowMessage} from './styles';
+import * as DividerHandlerManager from './dividerHandlerManager';
+
+type RenderedSpanTree = {
+  spanTree: JSX.Element | null;
+  nextSpanNumber: number;
+  numOfHiddenSpansAbove: number;
+};
+
+type PropType = {
+  traceViewRef: React.RefObject<HTMLDivElement>;
+  trace: ParsedTraceType;
+  dragProps: DragManagerChildrenProps;
+};
+
+class SpanTree extends React.Component<PropType> {
+  shouldComponentUpdate(nextProps: PropType) {
+    if (nextProps.dragProps.isDragging) {
+      return false;
+    }
+
+    return true;
+  }
+
+  renderSpan = ({
+    spanNumber,
+    treeDepth,
+    numOfHiddenSpansAbove,
+    spanID,
+    traceID,
+    lookup,
+    span,
+    generateBounds,
+    pickSpanBarColour,
+  }: {
+    spanNumber: number;
+    treeDepth: number;
+    numOfHiddenSpansAbove: number;
+    spanID: string;
+    traceID: string;
+    span: Readonly<SpanType>;
+    lookup: Readonly<SpanChildrenLookupType>;
+    generateBounds: (bounds: SpanBoundsType) => SpanGeneratedBoundsType;
+    pickSpanBarColour: () => string;
+  }): RenderedSpanTree => {
+    const spanBarColour: string = pickSpanBarColour();
+
+    const spanChildren: Array<SpanType> = get(lookup, spanID, []);
+
+    const bounds = generateBounds({
+      startTimestamp: span.start_timestamp,
+      endTimestamp: span.timestamp,
+    });
+
+    const isCurrentSpanHidden = !bounds.isSpanVisibleInView;
+
+    type AccType = {
+      renderedSpanChildren: Array<JSX.Element>;
+      nextSpanNumber: number;
+      numOfHiddenSpansAbove: number;
+    };
+
+    const reduced: AccType = spanChildren.reduce(
+      (acc: AccType, spanChild) => {
+        const key = `${traceID}${spanChild.span_id}`;
+
+        const results = this.renderSpan({
+          spanNumber: acc.nextSpanNumber,
+          treeDepth: treeDepth + 1,
+          numOfHiddenSpansAbove: acc.numOfHiddenSpansAbove,
+          span: spanChild,
+          spanID: spanChild.span_id,
+          traceID,
+          lookup,
+          generateBounds,
+          pickSpanBarColour,
```

Comment:
Couldn't the span figure this out by itself given the # of hidden spans above + depth + span number or some other information, rather than pass through the same function everywhere?

Reply:
done in https://github.com/getsentry/sentry/pull/14225/commits/e984adb8b98a5f88bae1afa73166a978ceadcddf 👍